routing/router.go

```diff
@@ -2249,3 +2249,288 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+				"constraints [%v - %v] violated with amt %v",
+				policy.MinHTLC, maxHtlc, amt)
+		}
+
+		// Update running amount by adding the fee for non-local
+		// channels.
+		if !localChan {
+			amt += policy.ComputeFee(amt)
+		}
+
+		return runningAmounts{
+			amt: amt,
+		}, nil
+	}
+
+	// Adapt the minimum amount to what this channel allows.
+	if policy.MinHTLC > amts.amt {
+		amt = policy.MinHTLC
+	}
+
+	// Update the maximum amount too to be able to detect incompatible
+	// channels.
+	max := amts.max
+	if maxHtlc < amts.max {
+		max = maxHtlc
+	}
+
+	// If we get in the situation that the minimum amount exceeds
+	// the maximum amount (enforced further down stream), we have
+	// incompatible channel policies.
+	//
+	// There is possibility with pubkey addressing that we should
+	// have selected a different channel downstream, but we don't
+	// backtrack to try to fix that.
+	if amt > max {
+		return runningAmounts{},
+			fmt.Errorf("incompatible channel policies: %v "+
+				"exceeds %v", amt, max)
+	}
+
+	// Add fees to the running amounts. Skip the source node fees as
+	// those do not need to be paid.
+	if !localChan {
+		amt += policy.ComputeFee(amt)
+		max += policy.ComputeFee(max)
+	}
+
+	return runningAmounts{amt: amt, max: max}, nil
+}
+
+// ErrNoChannel is returned when a route cannot be built because there are no
+// channels that satisfy all requirements.
+type ErrNoChannel struct {
+	position int
+	fromNode route.Vertex
+}
+
+// Error returns a human readable string describing the error.
+func (e ErrNoChannel) Error() string {
+	return fmt.Sprintf("no matching outgoing channel available for "+
+		"node %v (%v)", e.position, e.fromNode)
+}
+
+// BuildRoute returns a fully specified route based on a list of pubkeys. If
+// amount is nil, the minimum routable amount is used. To force a specific
+// outgoing channel, use the outgoingChan parameter.
+func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
+	hops []route.Vertex, outgoingChan *uint64,
+	finalCltvDelta int32) (*route.Route, error) {
+
+	log.Tracef("BuildRoute called: hopsCount=%v, amt=%v",
+		len(hops), amt)
+
+	// We'll attempt to obtain a set of bandwidth hints that helps us select
+	// the best outgoing channel to use in case no outgoing channel is set.
+	bandwidthHints, err := generateBandwidthHints(
+		r.selfNode, r.cfg.QueryBandwidth,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Allocate a list that will contain the selected channels for this
+	// route.
+	edges := make([]*channeldb.ChannelEdgePolicy, len(hops))
+
+	// Keep a running amount and the maximum for this route.
+	amts := runningAmounts{
+		max: lnwire.MilliSatoshi(^uint64(0)),
+	}
+	if amt != nil {
+		amts.amt = *amt
+	} else {
+		// For minimum amount routes, aim to deliver at least 1 msat to
+		// the destination. There are nodes in the wild that have a
+		// min_htlc channel policy of zero, which could lead to a zero
+		// amount payment being made.
+		amts.amt = 1
```

Comment:
:+1:

Reply:
Yes, that showed up during probing on mainnet.